In [6]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import os

In [7]:
df = pd.read_excel("c:\\Users\\shaba\\PycharmProjects\\Data-Analysis\\Data\Online Retail.xlsx")

<>:1: SyntaxWarning: invalid escape sequence '\O'
<>:1: SyntaxWarning: invalid escape sequence '\O'
C:\Users\shaba\AppData\Local\Temp\ipykernel_11084\3107174252.py:1: SyntaxWarning: invalid escape sequence '\O'
  df = pd.read_excel("c:\\Users\\shaba\\PycharmProjects\\Data-Analysis\\Data\Online Retail.xlsx")


In [8]:
df.sample(5)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
64877,541679,21218,RED SPOTTY BISCUIT TIN,6,2011-01-20 14:17:00,3.75,15422.0,United Kingdom
464097,576077,84559A,3D SHEET OF DOG STICKERS,1,2011-11-13 16:22:00,0.85,16720.0,United Kingdom
131514,547568,20914,SET/5 RED RETROSPOT LID GLASS BOWLS,6,2011-03-24 09:50:00,2.95,13341.0,United Kingdom
376124,569494,23327,HANGING CLEAR MINI BOTTLE,192,2011-10-04 13:42:00,0.55,15739.0,United Kingdom
407513,571890,23399,HOME SWEET HOME HANGING HEART,12,2011-10-19 14:05:00,0.85,16826.0,United Kingdom


In [9]:
df.shape

(541909, 8)

In [10]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [19]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [13]:
df.duplicated().sum()

np.int64(5268)

In [14]:
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386048,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


## Quality
- Missing values in Description and CustomerID
- CustomerID datatype is float and InvoiceDate datatye is datetime64[ns]
- Negative values in quantity
- inconsistency in InvoiceNo and StockCode i.e. contains alphabets at start in some InvoiceNo and at end In some StockCode 

## Tidiness

- Same column for date and time i.e. InvoiceDate
- 5268 duplicate rows

In [15]:
data = df.copy()

#### Handling Missing Values

In [31]:
print(f"missing_Description' = {(data.isnull().sum()['Description']/len(data)) *100:.2f} %")

missing_Description' = 0.27 %


In [32]:
print(f"missing_CustomerID = {(data.isnull().sum()['CustomerID']/len(data)) *100:.2f} %")

missing_CustomerID = 24.93 %


In [33]:
len(df['Description'].dropna())/ len(df)

0.9973168926886249

In [34]:
new_df = df['Description'].dropna()
df.shape, new_df.shape

((541909, 8), (540455,))

In [ ]:
plt.hist(df)